File: packages/spend-capsule-protocol/src/merkle.ts
import { sha256 } from "@noble/hashes/sha2";
import { bytesToHex, hexToBytes } from "@noble/hashes/utils";
import { canonicalize, sha256Prefixed } from "./hash.js";
import {
  type ChainVerifyResult,
  GENESIS_PREV_RECEIPT_HASH,
  type ReceiptPayload,
  RECEIPT_VERSION,
} from "./types.js";
import { ValidationError, validateReceiptPayload } from "./validate.js";
import { parseRfc3339Strict } from "./rfc3339.js";

// Domain-separation tags prevent cross-context collisions. Without these, a
// 32-byte value that happens to equal a leaf hash could be reinterpreted as
// an internal node digest (or vice versa) and alter the root. See
// https://en.wikipedia.org/wiki/Merkle_tree#Second_preimage_attack.
const DOMAIN_LEAF = 0x00;
const DOMAIN_NODE = 0x01;
const DOMAIN_ANCHOR = 0x02;
const DOMAIN_ROOT = 0x03;

export const MERKLE_BLOCK_SIZE = 1024;

// Genesis merkle root has its own domain context so no real receipt root can
// equal it structurally.
export const GENESIS_MERKLE_ROOT: string = (() => {
  const seed = new TextEncoder().encode("veto.merkle.genesis/1");
  const tagged = new Uint8Array(seed.length + 1);
  tagged[0] = DOMAIN_ANCHOR;
  tagged.set(seed, 1);
  return sha256Prefixed(tagged);
})();

export type ReceiptDraft = Omit<ReceiptPayload, "version" | "prev_receipt_hash" | "merkle_root">;

export function hashReceipt(receipt: ReceiptPayload): string {
  return sha256Prefixed(canonicalize(receipt));
}

function stripPrefix(h: string): Uint8Array {
  if (!h.startsWith("sha256:")) {
    throw new Error(`digest must be sha256:<hex>, got ${h.slice(0, 16)}...`);
  }
  const hex = h.slice(7);
  if (!/^[0-9a-f]{64}$/.test(hex)) {
    throw new Error(`digest must be exactly 64 lowercase hex chars; got length ${hex.length}`);
  }
  return hexToBytes(hex);
}

function taggedHash(domain: number, ...parts: Uint8Array[]): Uint8Array {
  const total = 1 + parts.reduce((n, p) => n + p.length, 0);
  const buf = new Uint8Array(total);
  buf[0] = domain;
  let off = 1;
  for (const p of parts) {
    buf.set(p, off);
    off += p.length;
  }
  return sha256(buf);
}

function u64Bytes(n: number): Uint8Array {
  if (!Number.isSafeInteger(n) || n < 0) {
    throw new Error(`u64Bytes requires a non-negative safe integer; got ${n}`);
  }
  const out = new Uint8Array(8);
  // Big-endian 64-bit encoding. Safe up to 2^53 which is far more than any
  // realistic chain length.
  for (let i = 7; i >= 0; i--) {
    out[i] = n & 0xff;
    n = Math.floor(n / 256);
  }
  return out;
}

/**
 * Compute a merkle root over leaf receipt hashes.
 *
 * Leaf count is bound into the root via a DOMAIN_ROOT wrapper. This closes
 * CVE-2012-2459-style duplicate-last ambiguity, where `[a, b, c]` and
 * `[a, b, c, c]` would otherwise produce identical inner roots and leave the
 * membership size uncommitted.
 */
export function computeMerkleRoot(leaves: string[]): string {
  if (leaves.length === 0) return GENESIS_MERKLE_ROOT;
  // Leaves are already sha256(canonical(receipt)); re-hash each under the
  // leaf domain so internal nodes and leaves cannot collide.
  let level = leaves.map((l) => taggedHash(DOMAIN_LEAF, stripPrefix(l)));
  while (level.length > 1) {
    const next: Uint8Array[] = [];
    for (let i = 0; i < level.length; i += 2) {
      const left = level[i]!;
      const right = i + 1 < level.length ? level[i + 1]! : left;
      next.push(taggedHash(DOMAIN_NODE, left, right));
    }
    level = next;
  }
  const inner = level[0]!;
  // Wrap inner root with explicit leaf count. Any two chains that differ in
  // size MUST produce a different root.
  const rootBytes = taggedHash(DOMAIN_ROOT, u64Bytes(leaves.length), inner);
  return `sha256:${bytesToHex(rootBytes)}`;
}

/**
 * Combine a previous rolling anchor with a newly-built block root into a new
 * rolling anchor. Bound under the anchor domain so the value cannot be
 * reinterpreted as a leaf or internal node.
 */
export function combineAnchors(prev: string, next: string): string {
  return `sha256:${bytesToHex(
    taggedHash(DOMAIN_ANCHOR, stripPrefix(prev), stripPrefix(next)),
  )}`;
}

export interface BuildReceiptInput {
  draft: ReceiptDraft;
  prev: ReceiptPayload | null;
  merkleRoot?: string;
}

export function buildReceipt(input: BuildReceiptInput): ReceiptPayload {
  const prevHash = input.prev ? hashReceipt(input.prev) : GENESIS_PREV_RECEIPT_HASH;
  const merkleRoot = input.merkleRoot ?? input.prev?.merkle_root ?? GENESIS_MERKLE_ROOT;
  const built: ReceiptPayload = {
    version: RECEIPT_VERSION,
    ...input.draft,
    prev_receipt_hash: prevHash,
    merkle_root: merkleRoot,
  };
  // Schema-validate on build so bad drafts fail loudly at the right layer.
  try {
    validateReceiptPayload(built);
  } catch (err) {
    if (err instanceof ValidationError) {
      throw new Error(`buildReceipt produced invalid receipt: ${err.message}`);
    }
    throw err;
  }
  return built;
}

export interface ChainVerifyOptions {
  /**
   * Maximum allowed backward skew when comparing consecutive `issued_at`
   * timestamps. Chains are expected to be monotonically non-decreasing;
   * minor clock drift within this window is tolerated (default: 5s).
   */
  timestampSkewSeconds?: number;
}

/**
 * Verify a per-entity receipt chain end-to-end. Enforces:
 *   1. Every receipt passes structural validation.
 *   2. `version` matches the expected protocol constant.
 *   3. `prev_receipt_hash` links to the previous receipt (or genesis at i=0).
 *   4. `issued_at` is non-decreasing (allows equal; rejects backward jumps
 *      beyond `timestampSkewSeconds`).
 *   5. `merkle_root` only changes at block boundaries — within a block it
 *      must match the previous receipt's root. This makes inserting a
 *      forged root mid-chain detectable without the verifier needing a
 *      separate anchor log.
 */
export function verifyReceiptChain(
  receipts: ReceiptPayload[],
  options: ChainVerifyOptions = {},
): ChainVerifyResult {
  if (receipts.length === 0) return { ok: true };
  const skewMs = (options.timestampSkewSeconds ?? 5) * 1000;

  let prevIssuedMs = -Infinity;
  let prevRoot: string | null = null;

  for (let i = 0; i < receipts.length; i++) {
    const r = receipts[i]!;

    // (1) Structural validation. A receipt that doesn't pass its own schema
    // is not "valid, just in a broken chain" — it's not a receipt at all.
    try {
      validateReceiptPayload(r);
    } catch (err) {
      const msg = err instanceof ValidationError ? err.message : String(err);
      return { ok: false, breakAt: i, reason: `receipt[${i}] invalid: ${msg}` };
    }

    // (2) Version.
    if (r.version !== RECEIPT_VERSION) {
      return {
        ok: false,
        breakAt: i,
        reason: `receipt[${i}] has unsupported version ${r.version}`,
      };
    }

    // (3) Hash-link continuity.
    const expectedPrev = i === 0 ? GENESIS_PREV_RECEIPT_HASH : hashReceipt(receipts[i - 1]!);
    if (r.prev_receipt_hash !== expectedPrev) {
      return {
        ok: false,
        breakAt: i,
        reason:
          i === 0
            ? `receipt[0] prev_receipt_hash must be the genesis hash`
            : `receipt[${i}] prev_receipt_hash does not match sha256 of receipt[${i - 1}]`,
      };
    }

    // (4) Monotonic issued_at.
    const issuedMs = parseRfc3339Strict(r.issued_at).epochMs;
    if (issuedMs + skewMs < prevIssuedMs) {
      return {
        ok: false,
        breakAt: i,
        reason: `receipt[${i}] issued_at ${r.issued_at} precedes receipt[${i - 1}].issued_at beyond tolerated skew`,
      };
    }
    prevIssuedMs = Math.max(prevIssuedMs, issuedMs);

    // (5) Merkle root progression.
    //
    // A merkle_root change mid-chain is only legitimate at a block boundary
    // (every MERKLE_BLOCK_SIZE receipts). Within a block the root is the
    // previous block's root — receipts commit to their latest anchor. A
    // silent root swap mid-block indicates tampering or an out-of-band
    // anchor insertion.
    if (prevRoot !== null && r.merkle_root !== prevRoot) {
      const atBoundary = i % MERKLE_BLOCK_SIZE === 0;
      if (!atBoundary) {
        return {
          ok: false,
          breakAt: i,
          reason: `receipt[${i}] merkle_root changed mid-block (position ${i % MERKLE_BLOCK_SIZE})`,
        };
      }
    }
    prevRoot = r.merkle_root;
  }
  return { ok: true };
}

export interface MerkleAnchor {
  entity_id: string;
  chain_index_start: number;
  chain_index_end: number;
  block_root: string;
  rolling_root: string;
  issued_at: string;
}

/**
 * Compute a rolling merkle anchor for a block of receipts. Binds
 * `entity_id`, the chain index range, and `issued_at` into the rolling root
 * so a forged `block_root` with the same bytes but a different label chain
 * cannot be spliced in.
 */
export function anchorBlock(
  entityId: string,
  startIndex: number,
  receipts: ReceiptPayload[],
  prevAnchor: MerkleAnchor | null,
  now: Date = new Date(),
): MerkleAnchor {
  if (receipts.length === 0) {
    throw new Error("anchorBlock requires at least one receipt");
  }
  if (!Number.isInteger(startIndex) || startIndex < 0) {
    throw new Error("anchorBlock startIndex must be a non-negative integer");
  }
  const blockRoot = computeMerkleRoot(receipts.map(hashReceipt));
  const issuedAt = now.toISOString();
  const endIndex = startIndex + receipts.length - 1;

  // Bind anchor identity into the rolling root. Without this, two chains with
  // identical block_roots but different entity_ids would produce identical
  // anchors.
  const boundInput = canonicalize({
    entity_id: entityId,
    chain_index_start: startIndex,
    chain_index_end: endIndex,
    block_root: blockRoot,
    issued_at: issuedAt,
  });
  const blockBound = sha256Prefixed(boundInput);

  const prevSeed = prevAnchor?.rolling_root ?? GENESIS_MERKLE_ROOT;
  const rollingRoot = combineAnchors(prevSeed, blockBound);

  return {
    entity_id: entityId,
    chain_index_start: startIndex,
    chain_index_end: endIndex,
    block_root: blockRoot,
    rolling_root: rollingRoot,
    issued_at: issuedAt,
  };
}

File: packages/sdk-python/veto/capsule/merkle.py
"""Receipt chain + merkle root primitives."""

from __future__ import annotations

import datetime as _dt
import hashlib
import re
from dataclasses import dataclass
from typing import Any

from .hash import canonicalize, sha256_prefixed
from .rfc3339 import parse_rfc3339_strict
from .types import (
    GENESIS_PREV_RECEIPT_HASH,
    RECEIPT_VERSION,
    ChainVerifyResult,
    ReceiptPayload,
)
from .validate import ValidationError, validate_receipt_payload

MERKLE_BLOCK_SIZE = 1024

# Domain tags — see merkle.ts for rationale. Without domain separation a
# 32-byte value that equals a leaf hash could be reinterpreted as an internal
# node digest and change the computed root.
_DOMAIN_LEAF = 0x00
_DOMAIN_NODE = 0x01
_DOMAIN_ANCHOR = 0x02
_DOMAIN_ROOT = 0x03


def _tagged_hash(domain: int, *parts: bytes) -> bytes:
    return hashlib.sha256(bytes([domain]) + b"".join(parts)).digest()


def _u64_bytes(n: int) -> bytes:
    if n < 0 or n > 2**64 - 1:
        raise ValueError(f"_u64_bytes requires 0 <= n < 2^64; got {n}")
    return n.to_bytes(8, "big")


GENESIS_MERKLE_ROOT: str = "sha256:" + _tagged_hash(
    _DOMAIN_ANCHOR, b"veto.merkle.genesis/1"
).hex()

_RE_SHA256_HEX64 = re.compile(r"^[0-9a-f]{64}$")


def hash_receipt(receipt: ReceiptPayload) -> str:
    return sha256_prefixed(canonicalize(receipt))


def _strip_prefix(h: str) -> bytes:
    if not h.startswith("sha256:"):
        raise ValueError(f"digest must be sha256:<hex>, got {h[:16]}...")
    hex_str = h[7:]
    if not _RE_SHA256_HEX64.match(hex_str):
        raise ValueError(
            f"digest must be exactly 64 lowercase hex chars; got length {len(hex_str)}"
        )
    return bytes.fromhex(hex_str)


def compute_merkle_root(leaves: list[str]) -> str:
    """Compute merkle root with leaf count bound into the output.

    The DOMAIN_ROOT wrapper closes CVE-2012-2459-style duplicate-last
    ambiguity: `[a, b, c]` and `[a, b, c, c]` would otherwise produce
    identical inner roots.
    """
    if not leaves:
        return GENESIS_MERKLE_ROOT
    level = [_tagged_hash(_DOMAIN_LEAF, _strip_prefix(leaf)) for leaf in leaves]
    while len(level) > 1:
        nxt: list[bytes] = []
        for i in range(0, len(level), 2):
            left = level[i]
            right = level[i + 1] if i + 1 < len(level) else left
            nxt.append(_tagged_hash(_DOMAIN_NODE, left, right))
        level = nxt
    inner = level[0]
    root = _tagged_hash(_DOMAIN_ROOT, _u64_bytes(len(leaves)), inner)
    return "sha256:" + root.hex()


def combine_anchors(prev: str, nxt: str) -> str:
    return "sha256:" + _tagged_hash(
        _DOMAIN_ANCHOR, _strip_prefix(prev), _strip_prefix(nxt)
    ).hex()


def build_receipt(
    draft: dict[str, Any],
    prev: ReceiptPayload | None,
    *,
    merkle_root: str | None = None,
) -> ReceiptPayload:
    """Build a receipt, setting version, prev_receipt_hash, and merkle_root.

    Schema-validates the result so bad drafts fail loudly.
    """
    prev_hash = hash_receipt(prev) if prev else GENESIS_PREV_RECEIPT_HASH
    resolved_root = (
        merkle_root
        if merkle_root is not None
        else (prev["merkle_root"] if prev else GENESIS_MERKLE_ROOT)
    )
    receipt: ReceiptPayload = {
        "version": RECEIPT_VERSION,
        **draft,  # type: ignore[misc]
        "prev_receipt_hash": prev_hash,
        "merkle_root": resolved_root,
    }
    try:
        validate_receipt_payload(receipt)
    except ValidationError as err:
        raise ValueError(f"build_receipt produced invalid receipt: {err}") from err
    return receipt


def verify_receipt_chain(
    receipts: list[ReceiptPayload],
    *,
    timestamp_skew_seconds: int = 5,
) -> ChainVerifyResult:
    """Verify a per-entity receipt chain end-to-end.

    Enforces: structural validation, version, hash-link continuity,
    monotonic issued_at (tolerating `timestamp_skew_seconds` backward drift),
    and merkle_root progression (root only changes at block boundaries).
    """
    if not receipts:
        return {"ok": True}
    skew_ms = timestamp_skew_seconds * 1000
    prev_issued_ms: float = float("-inf")
    prev_root: str | None = None

    for i, r in enumerate(receipts):
        # (1) Structural validation.
        try:
            validate_receipt_payload(r)
        except ValidationError as err:
            return {"ok": False, "breakAt": i, "reason": f"receipt[{i}] invalid: {err}"}

        # (2) Version.
        if r.get("version") != RECEIPT_VERSION:
            return {
                "ok": False,
                "breakAt": i,
                "reason": f"receipt[{i}] has unsupported version {r.get('version')}",
            }

        # (3) Hash-link continuity.
        expected_prev = GENESIS_PREV_RECEIPT_HASH if i == 0 else hash_receipt(receipts[i - 1])
        if r["prev_receipt_hash"] != expected_prev:
            return {
                "ok": False,
                "breakAt": i,
                "reason": (
                    "receipt[0] prev_receipt_hash must be the genesis hash"
                    if i == 0
                    else f"receipt[{i}] prev_receipt_hash does not match sha256 of receipt[{i - 1}]"
                ),
            }

        # (4) Monotonic issued_at.
        issued_ms = parse_rfc3339_strict(r["issued_at"]).epoch_ms
        if issued_ms + skew_ms < prev_issued_ms:
            return {
                "ok": False,
                "breakAt": i,
                "reason": (
                    f"receipt[{i}] issued_at {r['issued_at']} precedes "
                    f"receipt[{i - 1}].issued_at beyond tolerated skew"
                ),
            }
        prev_issued_ms = max(prev_issued_ms, issued_ms)

        # (5) Merkle root progression.
        if prev_root is not None and r["merkle_root"] != prev_root:
            at_boundary = i % MERKLE_BLOCK_SIZE == 0
            if not at_boundary:
                return {
                    "ok": False,
                    "breakAt": i,
                    "reason": (
                        f"receipt[{i}] merkle_root changed mid-block "
                        f"(position {i % MERKLE_BLOCK_SIZE})"
                    ),
                }
        prev_root = r["merkle_root"]

    return {"ok": True}


@dataclass
class MerkleAnchor:
    entity_id: str
    chain_index_start: int
    chain_index_end: int
    block_root: str
    rolling_root: str
    issued_at: str


def anchor_block(
    entity_id: str,
    start_index: int,
    receipts: list[ReceiptPayload],
    prev_anchor: MerkleAnchor | None,
    *,
    now: _dt.datetime | None = None,
) -> MerkleAnchor:
    if not receipts:
        raise ValueError("anchor_block requires at least one receipt")
    if not isinstance(start_index, int) or start_index < 0:
        raise ValueError("anchor_block start_index must be a non-negative integer")

    now = now or _dt.datetime.now(_dt.timezone.utc)
    issued_at = now.isoformat().replace("+00:00", "Z")
    end_index = start_index + len(receipts) - 1
    block_root = compute_merkle_root([hash_receipt(r) for r in receipts])

    # Bind anchor identity into the rolling root. Without this, two chains
    # with identical block_roots under different entity_ids would produce
    # identical anchors — a replay/relabel attack.
    bound_input = canonicalize(
        {
            "entity_id": entity_id,
            "chain_index_start": start_index,
            "chain_index_end": end_index,
            "block_root": block_root,
            "issued_at": issued_at,
        }
    )
    block_bound = sha256_prefixed(bound_input)

    prev_seed = prev_anchor.rolling_root if prev_anchor else GENESIS_MERKLE_ROOT
    rolling_root = combine_anchors(prev_seed, block_bound)

    return MerkleAnchor(
        entity_id=entity_id,
        chain_index_start=start_index,
        chain_index_end=end_index,
        block_root=block_root,
        rolling_root=rolling_root,
        issued_at=issued_at,
    )
